examples/Nat/PacketEncapsulation.cs

```diff
@@ -0,0 +1,213 @@
+    /// </summary>
+    public EthernetPacket LinkPacket { get; }
+
+    /// <summary>
+    /// The network-layer packet.
+    /// </summary>
+    public IpPacket NetworkPacket { get; }
+
+    /// <summary>
+    /// The transport-layer packet.
+    /// </summary>
+    public TTransport TransportPacket { get; }
+
+    /// <summary>
+    /// Casts the packet and it's payload packets to the correct types.
+    /// </summary>
+    /// <param name="packet"></param>
+    public PacketEncapsulation(Packet packet)
+    {
+      LinkPacket = (EthernetPacket)packet;
+      NetworkPacket = (IpPacket)LinkPacket.PayloadPacket;
+      TransportPacket = (TTransport)NetworkPacket.PayloadPacket;
+    }
+
+    /// <summary>
+    /// Updates the checksums of the packets for each layer.
+    /// </summary>
+    public void UpdateChecksums()
+    {
+      // Update transport checksum
+      UpdateTransportChecksum();
+
+      // Update IPv4 checksum
+      if (NetworkPacket is IPv4Packet)
+        ((IPv4Packet)NetworkPacket).UpdateIPChecksum();
+    }
+
+    /// <summary>
+    /// Rewrites the packet's source.
+    /// </summary>
+    /// <param name="node">The source node.</param>
+    public void SetSource(TNode node)
+    {
+      // Rewrite the MAC address
+      LinkPacket.SourceHwAddress = node.MacAddress;
+
+      // Rewrite the source IP address
+      NetworkPacket.SourceAddress = node.Address;
+
+      // Allow subclasses to set the transport layer protocol values
+      OnSetSource(node);
+    }
+
+    /// <summary>
+    /// Rewrites the packet's destination.
+    /// </summary>
+    /// <param name="node">The destination node.</param>
+    public void SetDestination(TNode node)
+    {
+      // Rewrite the MAC address
+      LinkPacket.DestinationHwAddress = node.MacAddress;
+
+      // Rewrite the destination IP address
+      NetworkPacket.DestinationAddress = node.Address;
+
+      // Allow subclasses to set the transport layer protocol values
+      OnSetDestination(node);
+    }
+
+    /// <summary>
+    /// Gets the source node of this packet (where it appears to originate from).
+    /// </summary>
+    /// <param name="incomingNetworkInterface">The network interface this packet arrived on. Optional. If not supplied, Port_Drop is assumed.</param>
+    /// <returns>A <see cref="TNode"/> representing the source of the packet.</returns>
+    public abstract TNode GetSourceNode(int incomingNetworkInterface = NAT.Port_Drop);
+
+    /// <summary>
+    /// Gets the destination node of this packet.
+    /// </summary>
+    /// <param name="incomingNetworkInterface">The network interface this packet arrived on. Optional. If not supplied, Port_Drop is assumed.</param>
+    /// <returns>A <see cref="TNode"/> representing the source of the packet.</returns>
+    public abstract TNode GetDestinationNode(int incomingNetworkInterface = NAT.Port_Drop);
+
+    /// <summary>
+    /// This method is called whenever the SetSource method is called.
```

Comment:
It might be useful to have a short example explaining a situation when `SetSource` is called -- is it called frequently and in different circumstances, or rarely and in similar circumstances?